src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,205 @@
+                        {
+                            return;
+                        }
+
+                        // Field writes: assignment, increment/decrement or field passed by ref.
+                        var isFieldAssignemnt = fieldReference.Parent is IAssignmentOperation assignmentOperation &&
+                            assignmentOperation.Target == fieldReference;
+                        if (isFieldAssignemnt ||
+                            fieldReference.Parent is IIncrementOrDecrementOperation ||
+                            fieldReference.Parent is IArgumentOperation argumentOperation && argumentOperation.Parameter.RefKind != RefKind.None)
+                        {
+                            // Writes to fields inside constructor are ignored, except for the below cases:
+                            //  1. Instance reference of an instance field being written is not the instance being initialized by the constructor.
+                            //  2. Field is being written inside a lambda or local function.
+
+                            // Check if we are in the constructor of the containing type of the written field.
+                            if ((isConstructor || isStaticConstructor) &&
+                                field.ContainingType == operationBlockStartContext.OwningSymbol.ContainingType)
+                            {
+                                // For instance fields, ensure that the instance reference is being initialized by the constructor.
+                                var instanceFieldWrittenInCtor = isConstructor &&
+                                    fieldReference.Instance?.Kind == OperationKind.InstanceReference &&
+                                    (!isFieldAssignemnt || fieldReference.Parent.Parent?.Kind != OperationKind.ObjectOrCollectionInitializer);
+                                
+                                // For static fields, ensure that we are in the static constructor.
+                                var staticFieldWrittenInStaticCtor = isStaticConstructor && field.IsStatic;
+
+                                if (instanceFieldWrittenInCtor || staticFieldWrittenInStaticCtor)
+                                {
+                                    // Finally, ensure that the write is not inside a lambda or local function.
+                                    if (!IsInAnonymousFunctionOrLocalFunction(fieldReference))
+                                    {
+                                        // It is safe to ignore this write.
+                                        return;
+                                    }
+                                }
+                            }
+
+                            onFieldWrite(field);
+                        }
+                    }, OperationKind.FieldReference);
+                });
+
+                compilationStartContext.RegisterCompilationEndAction(compilationEndContext =>
```

Comment:
❕ Are compilation end actions already supported in the IDE layer?

* If so, refactor this callback to be a method (not local function) named `HandleCompilationEnd`
* If not, the analyzer needs to be refactored to avoid relying on this unsupported method

Reply:
> Are compilation end actions already supported in the IDE layer?

I don't think this is relevant here, though @heejaechang is planning to fix the compilation end action bug soon. This problem cannot be solved with local analysis without introducing performance problems or writing spaghetti code. Detection of unused fields, readonly fields, etc. should be a compilation analyzer. Presence of unrelated bugs in the IDE engine should not affect this analyzer's implementation logic.

Reply:
> I don't think this is relevant here, though @heejaechang is planning to fix the compilation end action bug soon. This problem cannot be solved with local analysis without introducing performance problems or writing spaghetti code. 

I don't think it's spaghetti code.  It's simply analysis of a symbol, which shouldn't have a perf problem.  

> Presence of unrelated bugs in the IDE engine should not affect this analyzer's implementation logic.

That's ok *only* if the fix for compilation end actions truly is coming soon.  We literally had features not on *for years* because we wrote them, expecting this part of the analysis engine to work, and it was never fixed.  So, i'm skeptical of any approach that depends on this until we actually fixed the underlying IDE analysis problem.